File: managers/src/main/go/com.github/controllers/category_controller.go
package controllers

import (
	"managers/src/main/go/com.github/dto"
	"managers/src/main/go/com.github/services"
)

type CategoryController struct {
	categoryService *services.CategoryService
}

func NewCategoryController(categoryService *services.CategoryService) *CategoryController {
	return &CategoryController{categoryService: categoryService}
}

// FindByCategoryName godoc
// @Summary Get category by category name
// @Description Get category by category name
// @Tags products category
// @Accept  json
// @Produce  json
// @Param categoryName path string true "Category Name"
// @Success 200 {object} dto.CategoryDto
// @Failure 400
// @Failure 403
// @Router /v1/category/{categoryName} [get]
func (controller *CategoryController) FindByCategoryName(categoryName string) (*dto.CategoryDto, error) {
	return controller.categoryService.ReadByName(categoryName)
}

// SaveCategory godoc
// @Summary Save category
// @Description Save product category
// @Tags products category
// @Accept  json
// @Produce  json
// @Param account body dto.CategoryDto true "Create category"
// @Success 200 {object} dto.CategoryDto
// @Failure 400
// @Failure 403
// @Router /v1/category [post]
func (controller *CategoryController) SaveCategory(payload *dto.CategoryDto) (*dto.CategoryDto, error) {
	return controller.categoryService.CreateCategory(payload)
}

// UpdateCategory godoc
// @Summary Update Category
// @Description Update category
// @Tags products category
// @Accept  json
// @Produce  json
// @Param account body dto.CategoryDto true "Create category"
// @Success 200
// @Failure 403
// @Failure 404
// @Router /v1/category [put]
func (controller *CategoryController) UpdateCategory(payload *dto.CategoryDto) error {
	return controller.categoryService.UpdateCategory(payload)
}

File: managers/src/main/java/com/github/managers/controllers/impl/CategoryController.java
package com.github.managers.controllers.impl;

import com.github.managers.controllers.ICategoryController;
import com.github.managers.dto.CategoryDto;
import com.github.managers.exceptions.NotFound;
import com.github.managers.services.ICategoryService;
import com.github.managers.utils.Logging;
import com.netflix.hystrix.contrib.javanica.annotation.HystrixCommand;
import lombok.RequiredArgsConstructor;
import org.springframework.util.StringUtils;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;

@RestController
@RequestMapping(path = "/v1/category")
@RequiredArgsConstructor
public class CategoryController implements ICategoryController {

    private final ICategoryService categoryService;

    @Override
    @HystrixCommand
    @Logging(isTime = true, isReturn = false)
    public CategoryDto saveCategory(@Valid CategoryDto payload) {
        return this.categoryService.create(payload)
                .orElseThrow(NotFound::new);
    }

    @Override
    @HystrixCommand
    @Logging(isTime = true, isReturn = false)
    public Object findByName(String name) {
        if (!StringUtils.hasText(name)) {
            return this.categoryService.readAll().orElseThrow(NotFound::new);
        }
        return this.categoryService.readByName(name)
                .orElseThrow(NotFound::new);
    }

    @Override
    @HystrixCommand
    @Logging(isTime = true, isReturn = false)
    public void updateCategory(@Valid CategoryDto payload) {
        this.categoryService.update(payload);
    }

    @Override
    @HystrixCommand
    @Logging(isTime = true, isReturn = false)
    public void removeCategory(Long id) {
        this.categoryService.delete(id);
    }
}
